pkg/server/middlewares.go

```diff
@@ -0,0 +1,98 @@
+package server
+
+import (
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	context "golang.org/x/net/context"
+
+	"github.com/luizalabs/teresa-api/models/storage"
+	"github.com/luizalabs/teresa-api/pkg/server/auth"
+	"github.com/luizalabs/teresa-api/pkg/server/teresa_errors"
+	"github.com/luizalabs/teresa-api/pkg/server/user"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+type serverStreamWrapper struct {
+	grpc.ServerStream
+	ctx context.Context
+}
+
+func (w *serverStreamWrapper) Context() context.Context {
+	return w.ctx
+}
+
+func loginStreamInterceptor(a auth.Auth, uOps user.Operations) grpc.StreamServerInterceptor {
+	return func(srv interface{}, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		if strings.HasSuffix(info.FullMethod, "Login") {
+			return handler(srv, stream)
+		}
+
+		ctx := stream.Context()
+		user, err := authorize(ctx, a, uOps)
+		if err != nil {
+			return err
+		}
+
+		ctx = context.WithValue(ctx, "user", user)
+		wrap := &serverStreamWrapper{stream, ctx}
+		return handler(srv, wrap)
+	}
+}
+
+func loginUnaryInterceptor(a auth.Auth, uOps user.Operations) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		if strings.HasSuffix(info.FullMethod, "Login") {
+			return handler(ctx, req)
+		}
+
+		user, err := authorize(ctx, a, uOps)
+		if err != nil {
+			return nil, err
+		}
+
+		ctx = context.WithValue(ctx, "user", user)
+		return handler(ctx, req)
+	}
+}
+
+func authorize(ctx context.Context, a auth.Auth, uOps user.Operations) (*storage.User, error) {
+	md, ok := metadata.FromContext(ctx)
+	if !ok {
+		return nil, auth.ErrPermissionDenied
+	}
+	if len(md["token"]) < 1 || md["token"][0] == "" {
+		return nil, auth.ErrPermissionDenied
+	}
+	email, err := a.ValidateToken(md["token"][0])
+	if err != nil {
+		return nil, err
+	}
+	return uOps.GetUser(email)
+}
+
+func recFunc(p interface{}) (err error) {
+	log.WithField("panic", p).Error("teresa-server recovered")
+	return status.Errorf(codes.Unknown, "Internal Server Error")
+}
+
+func logUnaryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	resp, err := handler(ctx, req)
+	if err != nil {
+		log.WithField("route", info.FullMethod).WithField("request", req).Error(err)
+		return resp, teresa_errors.Get(err)
+	}
+	return resp, nil
+}
+
+func logStreamInterceptor(srv interface{}, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+	err := handler(srv, stream)
+	if err != nil {
+		log.WithField("route", info.FullMethod).Error(err)
+		return teresa_errors.Get(err)
+	}
+	return nil
+}
```

Comment:
Also, I added the user of the request. Check some examples below and let know if you liked.

`time="2017-07-18T14:32:31Z" level=error msg="Log Interceptor got an Error" error="namespaces \"hashpass2\" already exists" request=name:"hashpass2" team:"arquitetura" limits:<default:<quantity:"500m" resource:"cpu" > default:<quantity:"512Mi" resource:"memory" > default_request:<quantity:"200m" resource:"cpu" > default_request:<quantity:"512Mi" resource:"memory" > > auto_scale:<cpu_target_utilization:70 max:2 min:1 >  route="/app.App/Create" user="diego.garcia@luizalabs.com"`

`time="2017-07-18T14:33:17Z" level=error msg="Log Interceptor got an Error" error="rpc error: code = NotFound desc = App not found: get annotation failed: namespaces \"hashpass3\" not found" route="/deploy.Deploy/Make" user="diego.garcia@luizalabs.com"`

ps: In the unary interceptor, I'm logging request content.